design/fast_finality.md

```diff
@@ -0,0 +1,276 @@
+
+* *Light batch*: A set of transactions produced by the designated sequencer, considered to be soft-finalised in that 
+  order
+
+## Requirements
+
+* Finality
+  * Transaction *soft* finality (finality guaranteed by the sequencer) is achieved in under one second
+  * There is eventual transaction *hard* finality (finality guaranteed by the L1)
+  * The sequencer is incentivised to achieve hard finality on an agreed cadence
+  * The sequencer is incentivised to hard-finalise transactions in the same order they are soft-finalised
+  * The sequencer is not able to "rewrite history" (or is strongly disincentivised from doing so), even for soft-final
+    transactions
+* Censorship resistance
+  * End-users can bypass the sequencer to include transactions on the L1 directly (possibly at higher cost and slower 
+    finality)
+  * The sequencer distributes all soft-finalised transactions to all nodes
+  * Nodes can easily prove whether a given soft-finalised transaction was hard-finalised, and in the correct order
+* Value-extraction resistance
+  * The sequencer cannot precompute the effects of running a given set of transactions without committing to that set 
+    of transactions (e.g. by running a single transaction, then `eth_call`ing to see how it has affected a smart 
+    contract's state)
+* Cost
+  * L1 transaction costs can be driven lower at the expense of extending the hard-finality window
+* User/dev experience
+  * The responses to RPC calls reflect the soft-finalised transactions, and not just the hard-finalised transactions
+* Resilience
+  * Failover does not require a governance action
+  * During failover, it is acceptable to break the one-second soft-finality guarantee 
+  * During failover, it is acceptable to drop transactions that have not been soft-finalised
+
+## Assumptions
+
+* There is a single aggregator that is also the genesis node. This node is known as the *sequencer*
+* The enclave has a start-up delay that exceeds the production time for light batches (see below)
+
+## Constraints
+
+* It must be possible to only publish rollups to the L1 every `x` blocks, where `x` >> 1
+* It must be possible to produce light batches (see below) at a higher frequency than L1 blocks
+
+## Design
+
+### Sequencer identity
+
+The sequencer's identity is given in the management contract on the L1 as a set of enclave attestations. This allows 
+other nodes to verify that the light batches and rollups are created by the sequencer. We use a set instead of a 
+single attestation to allow faster failover in the case of a sequencer node crashing (see the section `Resilience`, 
+below).
+
+Each attestation matches one of the sequencer's enclaves, and contains the hash of that enclave's key. Since the
+attestations are unique per machine, the enclaves cannot be impersonated. The foundation admin will then whitelist
+these attestations. If one of the sequencer's enclaves goes down irrecoverably, a replacement attestation is added and
```

Comment:
Got a bit confused here, you talk about the sequencer having a set of attestations already so not sure why replacement attestation is added at recovery time. But guess this bit is still being iterated on.

Reply:
I tried to clarify. The idea is that if you have a cluster of `n` enclaves, and one machine gets trashed, you have to add the attestation of a new machine, otherwise the cluster will shrink permanently.